File: lib/features/billers/data/biller_repository.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../../core/mock/mock_network.dart';
import 'biller.dart';
import 'biller_catalog.dart';
import 'catalog_strings.dart';

abstract interface class BillerRepository {
  Future<BillerCatalog> fetchCatalog(String language);
  Future<List<Biller>> search(String query, String language);
}

class FakeBillerRepository implements BillerRepository {
  FakeBillerRepository(this._network);

  final MockNetwork _network;

  static const _openAmountCategories = {'mobile-prepaid', 'dth', 'fastag'};
  static const _prefixIds = ['national', 'metro', 'city', 'state'];

  static final _catalogs = <String, BillerCatalog>{};

  BillerCatalog _catalogFor(String language) {
    return _catalogs.putIfAbsent(language, () => _buildCatalog(language));
  }

  static BillerCatalog _buildCatalog(String language) {
    final categories = categoryNames.entries
        .map((e) => BillerCategory(id: e.key, name: e.value[language]!))
        .toList();
    final billers = [
      for (final categoryId in categoryNames.keys)
        for (final prefixId in _prefixIds)
          Biller(
            id: '$categoryId-$prefixId',
            categoryId: categoryId,
            name: billerName(prefixId, categoryId, language),
            mode: _openAmountCategories.contains(categoryId)
                ? BillerMode.openAmount
                : BillerMode.presentment,
            inputParams: _paramsFor(categoryId, language),
          ),
    ];
    return BillerCatalog(categories: categories, billers: billers);
  }

  static List<BillerInputParam> _paramsFor(String categoryId, String language) {
    if (categoryId == 'credit-card') {
      return [
        BillerInputParam(
          key: 'card',
          label: paramLabels['Card Number']![language]!,
          hint: paramHints['Last 4 digits']![language]!,
        ),
        BillerInputParam(
          key: 'mobile',
          label: paramLabels['Registered Mobile']![language]!,
          hint: paramHints['10-digit mobile']![language]!,
        ),
      ];
    }
    final labelKey = switch (categoryId) {
      'mobile-postpaid' || 'mobile-prepaid' => 'Mobile Number',
      'dth' => 'Subscriber ID',
      'fastag' => 'Vehicle Number',
      'lpg' => 'LPG ID',
      'insurance' => 'Policy Number',
      'loan-emi' => 'Loan Account Number',
      'education' => 'Student ID',
      _ => 'Consumer Number',
    };
    final label = paramLabels[labelKey]![language]!;
    return [
      BillerInputParam(
        key: 'account',
        label: label,
        hint: enterHint(label, language),
      ),
    ];
  }

  @override
  Future<BillerCatalog> fetchCatalog(String language) async {
    await _network.delay();
    return _catalogFor(language);
  }

  @override
  Future<List<Biller>> search(String query, String language) async {
    await _network.delay();
    final needle = query.trim().toLowerCase();
    return _catalogFor(language)
        .billers
        .where((b) => b.name.toLowerCase().contains(needle))
        .toList();
  }
}

final billerRepositoryProvider = Provider<BillerRepository>(
  (ref) => FakeBillerRepository(ref.watch(mockNetworkProvider)),
);

File: src/features/billers/data/billerRepository.ts
import { MockNetwork } from '@/core/mock/mockNetwork';
import type { SupportedLanguage } from '@/core/format/localeConfigs';
import type { Biller, BillerCatalog, BillerInputParam } from './biller';
import { billerName, categoryNames, enterHint, paramHints, paramLabels } from './catalogStrings';

const openAmountCategories = new Set(['mobile-prepaid', 'dth', 'fastag']);
const prefixIds = ['national', 'metro', 'city', 'state'];

export interface BillerRepository {
  fetchCatalog(language: SupportedLanguage): Promise<BillerCatalog>;
  search(query: string, language: SupportedLanguage): Promise<Biller[]>;
}

function paramsFor(categoryId: string, language: SupportedLanguage): BillerInputParam[] {
  if (categoryId === 'credit-card') {
    return [
      { key: 'card', label: paramLabels['Card Number']![language], hint: paramHints['Last 4 digits']![language] },
      { key: 'mobile', label: paramLabels['Registered Mobile']![language], hint: paramHints['10-digit mobile']![language] },
    ];
  }
  const labelKey =
    categoryId === 'mobile-postpaid' || categoryId === 'mobile-prepaid' ? 'Mobile Number'
    : categoryId === 'dth' ? 'Subscriber ID'
    : categoryId === 'fastag' ? 'Vehicle Number'
    : categoryId === 'lpg' ? 'LPG ID'
    : categoryId === 'insurance' ? 'Policy Number'
    : categoryId === 'loan-emi' ? 'Loan Account Number'
    : categoryId === 'education' ? 'Student ID'
    : 'Consumer Number';
  const label = paramLabels[labelKey]![language];
  return [{ key: 'account', label, hint: enterHint(label, language) }];
}

function buildCatalog(language: SupportedLanguage): BillerCatalog {
  const categories = Object.entries(categoryNames).map(([id, names]) => ({ id, name: names[language] }));
  const billers: Biller[] = [];
  for (const categoryId of Object.keys(categoryNames)) {
    for (const prefixId of prefixIds) {
      billers.push({
        id: `${categoryId}-${prefixId}`,
        categoryId,
        name: billerName(prefixId, categoryId, language),
        mode: openAmountCategories.has(categoryId) ? 'openAmount' : 'presentment',
        inputParams: paramsFor(categoryId, language),
      });
    }
  }
  return { categories, billers };
}

export class FakeBillerRepository implements BillerRepository {
  private readonly catalogs = new Map<SupportedLanguage, BillerCatalog>();

  constructor(private readonly network: MockNetwork) {}

  private catalogFor(language: SupportedLanguage): BillerCatalog {
    const cached = this.catalogs.get(language);
    if (cached) return cached;
    const built = buildCatalog(language);
    this.catalogs.set(language, built);
    return built;
  }

  async fetchCatalog(language: SupportedLanguage): Promise<BillerCatalog> {
    await this.network.delay();
    return this.catalogFor(language);
  }

  async search(query: string, language: SupportedLanguage): Promise<Biller[]> {
    await this.network.delay();
    const needle = query.trim().toLowerCase();
    return this.catalogFor(language).billers.filter((b) => b.name.toLowerCase().includes(needle));
  }
}
